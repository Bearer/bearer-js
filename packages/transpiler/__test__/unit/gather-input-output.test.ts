import * as path from 'path'
import * as ts from 'typescript'
import * as TJS from 'typescript-json-schema'

import Metadata from '../../src/metadata'
import GatherIO from '../../src/transformers/gather-input-output'
const fixtures = path.join(__dirname, '..', '__fixtures__')

describe('GatherIO transformer', () => {
  describe('a simple root component', () => {
    it('has inputs and outputs added', () => {
      const metadata = new Metadata()
      metadata.registerComponent({
        classname: 'FeatDisplayComponent',
        fileName: '/__test__/__fixtures__/gather-input-output/root-component-transformed.tsx',
        finalTagName: 'complex-feature-display',
        name: 'complex',
        initialTagName: 'feat-display',
        isRoot: true
      })

      transpileFile('root-component-transformed.tsx', metadata)
      expect(metadata).toMatchSnapshot()
    })
  })
})

function transpileFile(filename: string, metadata: Metadata) {
  const files = [path.join(fixtures, 'gather-input-output', filename)]
  const config = ts.readConfigFile(
    path.join(__dirname, '../../../cli/templates/start', 'tsconfig.json'),
    ts.sys.readFile
  )
  const compilerHost = ts.createCompilerHost({})
  const programGenerator = TJS.getProgramFromFiles(files, config.config.compilerOptions, './ok')

  const generator = TJS.buildGenerator(programGenerator, {
    required: true
  })
  const program = ts.createProgram(
    files,
    { experimentalDecorators: true, outDir: path.join(__dirname, '../../.build-transformers') },
    compilerHost
  )
  program.emit(undefined, undefined, undefined, undefined, {
    before: [GatherIO({ metadata, generator })]
  })
}
