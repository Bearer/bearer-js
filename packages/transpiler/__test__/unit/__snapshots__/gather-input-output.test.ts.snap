// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GatherIO transformer a simple root component has inputs and outputs added 1`] = `
Metadata {
  "components": Array [
    Object {
      "classname": "FeatDisplayComponent",
      "fileName": "/__test__/__fixtures__/gather-input-output/root-component-transformed.tsx",
      "finalTagName": "complex-feature-display",
      "group": "complex",
      "initialTagName": "feat-display",
      "inputs": Array [
        Object {
          "default": undefined,
          "name": "noInitializer",
          "type": "string",
        },
        Object {
          "default": "'ok'",
          "name": "aStringProp",
          "type": "string",
        },
        Object {
          "default": "5",
          "name": "aNumberProp",
          "type": "number",
        },
      ],
      "isRoot": true,
      "outputs": Array [
        Object {
          "eventName": "bearer-sponge-bob-complex-namedevent",
          "name": "namedEvent",
          "payloadFormat": "unspecified",
        },
        Object {
          "eventName": "bearer-sponge-bob-complex-nontypedevent",
          "name": "nonTypedEvent",
          "payloadFormat": "unspecified",
        },
        Object {
          "eventName": "bearer-sponge-bob-complex-typedevent",
          "name": "typedEvent",
          "payloadFormat": Object {
            "properties": Object {
              "aNumber": Object {
                "description": "aNumber",
                "name": "aNumber",
                "required": true,
                "schema": Object {
                  "type": "number",
                },
              },
              "anEnum": Object {
                "description": "anEnum",
                "name": "anEnum",
                "required": true,
                "schema": Object {
                  "type": "object",
                },
              },
              "pullRequest": Object {
                "description": "pullRequest",
                "name": "pullRequest",
                "required": true,
                "schema": Object {
                  "$schema": "http://json-schema.org/draft-07/schema#",
                  "properties": Object {
                    "name": Object {
                      "type": "string",
                    },
                    "status": Object {
                      "enum": Array [
                        "closed",
                        "open",
                      ],
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "name",
                    "status",
                  ],
                  "type": "object",
                },
              },
            },
            "type": "object",
          },
        },
        Object {
          "eventName": "bearer-sponge-bob-complex-typedeventwithtype",
          "name": "typedEventWithType",
          "payloadFormat": Object {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "properties": Object {
              "anEnum": Object {
                "enum": Array [
                  "ko",
                  "ok",
                ],
                "type": "string",
              },
              "something": Object {
                "properties": Object {
                  "name": Object {
                    "type": "string",
                  },
                  "status": Object {
                    "enum": Array [
                      "closed",
                      "open",
                    ],
                    "type": "string",
                  },
                },
                "required": Array [
                  "name",
                  "status",
                ],
                "type": "object",
              },
              "somethingElse": Object {
                "type": "string",
              },
            },
            "required": Array [
              "anEnum",
              "something",
              "somethingElse",
            ],
            "type": "object",
          },
        },
        Object {
          "eventName": "bearer-sponge-bob-complex-typedeventwithtypearray",
          "name": "typedEventWithTypeArray",
          "payloadFormat": Object {
            "type": "object",
          },
        },
        Object {
          "eventName": "bearer-sponge-bob-complex-typedeventwithtypestring",
          "name": "typedEventWithTypeString",
          "payloadFormat": Object {
            "type": "string",
          },
        },
        Object {
          "eventName": "bearer-sponge-bob-complex-typedeventwithtypenumber",
          "name": "typedEventWithTypeNumber",
          "payloadFormat": Object {
            "type": "number",
          },
        },
        Object {
          "eventName": "bearer-sponge-bob-complex-typedeventwithtypeboolean",
          "name": "typedEventWithTypeBoolean",
          "payloadFormat": Object {
            "type": "boolean",
          },
        },
      ],
    },
  ],
  "prefix": null,
  "registerComponent": [Function],
  "suffix": null,
}
`;
